File: CS50/Week2/Readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

// delcare our functions
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int coleman_liau(int letters, int words, int sentences);

int main(void)
{
    // Get user promt fopr text
    string user_text = get_string("Text: ");

    // call letter counter function
    int letter_counts = count_letters(user_text);
    // printf("%i letters\n", letter_counts); // Test letter counts

    // call word counter function
    int word_counts = count_words(user_text);
    // printf("%i words\n", word_counts); // Test word counts

    // call our sentence counter function
    int sentence_counts = count_sentences(user_text);
    // printf("%i sentences\n", sentence_counts); // Test sentence counts

    // calculate Coleman-Liau algorithm. call function
    int grade = coleman_liau(letter_counts, word_counts, sentence_counts);

    // print grade based on conditions
    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade <= 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }

}

// Functions

// count the number of letters
int count_letters(string text)
{
    // make a letter counter
    int letters = 0;
    // Loop through each char in the text
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        // set up counts only if the letter are between a-z or A-Z
        if (isalpha(text[i]))
        {
            // add one to letter Count if condition is met
            letters += 1;
        }
    }
    return letters;
}

//  Count the number of words
int count_words(string text)
{
    // make a word count
    int word_count = 0;
    // loop through all of text
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        //  need to count the words only. If we count teh spaces that seperate words we will get all teh words
        // except teh last one. just add 1
        if (isspace(text[i]))
        {
            word_count += 1;
        }
    }
    return word_count + 1;
}

// Count number of sentences

int count_sentences(string text)
{
    // sentence counter
    int sentence_count = 0;
    //  iterate through our user text
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        //  loop for ending punctuation to count each sentance: '.' , '!' , or '?'
        if (text[i] == '.' | text[i] == '!' | text[i] == '?')
        {
            sentence_count += 1;
        }
    }
    return sentence_count;
}

// calculate formula
int coleman_liau(int letters, int words, int sentences)
{
    // concer ints to floats
    float L, S, idx;
    // L = get avergage number of letter per 100 words. Casted to float
    L = ((float) letters / words) * 100;

    // printf("L is = to: %f\n", L); //check L

    // S = get average number of sentences per 100 words. Casted to float
    S = ((float) sentences / words) * 100;

    // printf("S is = to: %f\n", S); // check  S

    // Coleman-Liau index algorithm: index(grade) = 0.0588 * L - 0.296 * S -15.8
    idx = round((0.0588 * L) - (0.296 * S) - 15.8); // needed to round as well

    // printf("Grade is = to: %f", idx); // check grade


    return idx;

}

File: CS50/Week6/readability.py
# Figure out grade level

punctuations = [".", "!", "?"]


def main():
    # get user input
    text = input("Enter some text to see grade level: ")

    # get letter counts
    letters = letter_counter(text)

    # get sentence counts
    sentences = sentence_counter(text)

    # get word counts
    words = word_counter(text)

    # calculate Coleman-Liau index
    Coleman_Liau = Coleman_Liau_index(letters, sentences, words)

    if Coleman_Liau < 1:
        print("Before Grade 1")
    elif Coleman_Liau > 16:
        print("Grade 16+")
    else:
        print(f"Grade {Coleman_Liau}")


def letter_counter(text):
    """Count number of letters"""
    letters = 0
    for char in text:
        if char.isalpha():
            letters += 1
    return letters


def sentence_counter(text):
    "Count number of sentences based on puncuation"""
    sentences = 0
    for char in text:
        if char in punctuations:
            sentences += 1
    return sentences


def word_counter(text):
    """count teh number of words"""
    words = 0
    for char in text:
        if char == " ":
            words += 1
    return words + 1


def Coleman_Liau_index(letters, sentences, words):
    """Coleman-Liau index is computed as 0.0588 * L - 0.296 * S - 15.8, where L is the average number of letters per 100 words in the text,
    #  and S is the average number of sentences per 100 words in the text."""
    L = (letters / words) * 100
    S = (sentences / words) * 100
    Coleman_Liau = round((0.0588 * L) - (0.296 * S) - 15.8)
    return Coleman_Liau


main()
